### ISSUE TITLE
Add support of symlink for Windows targets

### ISSUE THREAD

**Issue by GunGraveKoga** [Midar/objfw#36]
Since Windows Vista/Sever 2008 Win API has method CreateSymbolicLink()
I think that OFFileManager should to implement runtime detection of Windows major version and use kernel32.dll to load function "CreateSymbolicLink" dynamically
```objc
typedef BOOL WINAPI (*symlink_fn)(const wchar_t*, const wchar_t*, DWORD);
static bool __has_symbolic_link_support = false;
static symlink_fn create_symbolic_link = NULL;

+ (void)load 
{
	
	OSVERSIONINFOW osvi;
	ZeroMemory(&osvi, sizeof(OSVERSIONINFOW));
    osvi.dwOSVersionInfoSize = sizeof(OSVERSIONINFOW);

    GetVersionExW(&osvi);

    if (osvi.dwMajorVersion >= 6) {
    	__has_symbolic_link_support = true;

    	if (!create_symbolic_link) {
			create_symbolic_link = (symlink_fn) GetProcAddress(GetModuleHandle("kernel32.dll"), "CreateSymbolicLinkW");

			if (!create_symbolic_link)
				create_symbolic_link = (symlink_fn)-1;
		}
    }

}
```

**Comment by Midar** [comment]
Sounds like a good idea. Going to implement this. I suppose throwing OFNotImplementedException on older versions would be the best thing to do there.

**Comment by GunGraveKoga** [comment]
Remark:
As I see OFFileManager does not have unlink or similar method for deleting symbolic links, if ObjFW API offer to use removeItemAtPath: for remove symbolic links, then on Win32 this method should to use DeleteFileW and RemoveDirectoryW for files and folder accordingly

https://msdn.microsoft.com/en-us/library/windows/desktop/aa363866(v=vs.85).aspx

**Comment by Midar** [comment]
I added support for creating hardlinks and symlinks on Windows now.

I have not found a readlink() equivalent, though, to read the destination of a symlink.

**Comment by GunGraveKoga** [comment]
And what about "symbolicLinkExistsAtPath:" method?

I use something like this:
```objc
if (path == nil)
		@throw [OFInvalidArgumentException exception];

    if (__has_symbolic_link_support) {
    	WIN32_FIND_DATAW findData;
    	ZeroMemory(&findData, sizeof(WIN32_FIND_DATAW));
    	HANDLE h;

    	if ((h = FindFirstFileW((LPCWSTR)[path UTF16String], &findData)) == INVALID_HANDLE_VALUE)
    		return false;
    	else {
    		bool result = false;

    		if ((findData.dwFileAttributes & FILE_ATTRIBUTE_REPARSE_POINT) && (findData.dwReserved0 == IO_REPARSE_TAG_SYMLINK))
    			result = true;

    		FindClose(h);

    		return result;
    	}

    } else {
    	return false;
    }

    return false;
```
for not supported Windows version always return false.

For readlink() equivalent on Windows better to use low level API

```objc
- (OFString*)destinationOfSymbolicLinkAtPath:(OFString*)path
{
	if (path == nil)
		@throw [OFInvalidArgumentException exception];

    if (__has_symbolic_link_support) {

    	if (![self symbolicLinkExistsAtPath:path])
    		@throw [OFStatItemFailedException exceptionWithPath:path errNo:EINVAL];

    	HANDLE hLink = CreateFileW((LPCWSTR)[path UTF16String], GENERIC_READ, 
    		(FILE_SHARE_READ | FILE_SHARE_WRITE | FILE_SHARE_DELETE), NULL, OPEN_EXISTING, 
    		(FILE_FLAG_BACKUP_SEMANTICS | FILE_FLAG_OPEN_REPARSE_POINT), NULL);

    	if (hLink == INVALID_HANDLE_VALUE)
    		@throw [OFStatItemFailedException exceptionWithPath:path errNo:errno];

    	unsigned char buffer[MAXIMUM_REPARSE_DATA_BUFFER_SIZE];
		DWORD dummy = 0;

		if (DeviceIoControl(hLink, FSCTL_GET_REPARSE_POINT, NULL, 0, buffer, MAXIMUM_REPARSE_DATA_BUFFER_SIZE, &dummy, NULL)) {
			REPARSE_DATA_BUFFER *b = (REPARSE_DATA_BUFFER *) buffer;

			if (b->ReparseTag == IO_REPARSE_TAG_SYMLINK) {
				size_t len = b->SymbolicLinkReparseBuffer.SubstituteNameLength / sizeof(wchar_t);
				ssize_t offset = b->SymbolicLinkReparseBuffer.SubstituteNameOffset / sizeof(wchar_t);

				void* pool = objc_autoreleasePoolPush();

				OFString* result = [OFString stringWithUTF16String:(const of_char16_t *)(& b->SymbolicLinkReparseBuffer.PathBuffer[offset]) length:len];

				if ([result hasPrefix:@"\\??\\"])
					result = [result substringWithRange:of_range(4, [result length] - 4)];

				[result retain];

				objc_autoreleasePoolPop(pool);

				CloseHandle(hLink);

				return [result autorelease];
			}

			CloseHandle(hLink);

			@throw [OFStatItemFailedException exceptionWithPath:path errNo:EINVAL];
		}

    } else {
    	if (![self fileExistsAtPath:path] || ![self directoryExistsAtPath:path])
    		@throw [OFStatItemFailedException exceptionWithPath:path errNo:EINVAL];

    	return [[path copy] autorelease];
    }

    return nil;
}
```

**Comment by GunGraveKoga** [comment]
some macros like IO_REPARSE_TAG_SYMLINK may be not defined in headers for some Win32 targets like Win9x/NT, so it should be defined manually

```c
#ifndef IO_REPARSE_TAG_SYMLINK
#define IO_REPARSE_TAG_SYMLINK (__MSABI_LONG(0xA000000C))
#endif
```

**Comment by Midar** [comment]
Ok, let's reopen this then. Not really a fan of all that complexity, though. OTOH, CreateSymbolicLinkW was much shorter in the end, too ;).

**Comment by Midar** [comment]
I added `-[symbolicLinkExistsAtPath:]` and `-[destinationOfSymbolicLinkAtPath:]` now, since I could find everything for that on MSDN. It was spread across quite a number of pages, though, but I made it work :).

BTW, no need to provide entire code examples - pointing to the right functions is enough, I need to read it up on MSDN anyway and read through it very carefully. I also noticed that after reading MSDN, I came up with slightly different flags than you (some didn't too much sense).
